GeoLib/OctTree-impl.h

```diff
@@ -0,0 +1,260 @@
+		return false;
+	}
+
+	// the point pnt is not yet in the OctTree
+	if ((*pnt)[0] < _ll[0] || (*pnt)[1] < _ll[1] || (*pnt)[2] < _ll[2]) {
+		ret_pnt = nullptr;
+		return false;
+	}
+	if ((*pnt)[0] > _ur[0] || (*pnt)[1] > _ur[1] || (*pnt)[2] > _ur[2]) {
+		ret_pnt = nullptr;
+		return false;
+	}
+
+	// at this place it holds true that the point is within [_ll, _ur]
+	if (!_is_leaf) {
+		for (std::size_t k(0); k < 8; ++k) {
+			if (_childs[k]->addPoint_(pnt, ret_pnt)) {
+				return true;
+			} else {
+				if (ret_pnt != nullptr)
+					return false;
+			}
+		}
+	}
+
+	ret_pnt = pnt;
+
+	if (_pnts.size () < OctTree<POINT>::_max_items_per_node) {
+		_pnts.push_back(pnt);
+	} else { // i.e. _pnts.size () == OctTree<POINT>::_max_items_per_node
+		splitNode(pnt);
+		_pnts.clear();
+	}
+	return true;
+}
+
+template <typename POINT>
+template <typename T>
+void OctTree<POINT>::getPointsInRange(T const& min, T const& max,
+	std::vector<POINT*> &pnts) const
+{
+	if (_ur[0] < min[0]) return;
+	if (_ur[1] < min[1]) return;
+	if (_ur[2] < min[2]) return;
+
+	if (max[0] < _ll[0]) return;
+	if (max[1] < _ll[1]) return;
+	if (max[2] < _ll[2]) return;
+
+	if (_is_leaf) {
+		for (auto p : _pnts) {
+			if (min[0] <= (*p)[0] && (*p)[0] < max[0]
+				&& min[1] <= (*p)[1] && (*p)[1] < max[1]
+				&& min[2] <= (*p)[2] && (*p)[2] < max[2])
+				pnts.push_back(p);
+		}
+	} else {
+		for (std::size_t k(0); k<8; k++) {
+			_childs[k]->getPointsInRange(min, max, pnts);
+		}
+	}
+}
+
+template <typename POINT>
+OctTree<POINT>::OctTree(MathLib::Point3d const& ll, MathLib::Point3d const& ur)
+	: _ll(ll), _ur(ur), _is_leaf(true)
+{
+	// init childs
+	for (std::size_t k(0); k<8; ++k)
+		_childs[k] = nullptr;
+}
+
+template <typename POINT>
+bool OctTree<POINT>::addPoint_(POINT * pnt, POINT *& ret_pnt)
+{
+	if ((*pnt)[0] < _ll[0] || (*pnt)[1] < _ll[1] || (*pnt)[2] < _ll[2]) {
+		ret_pnt = nullptr;
+		return false;
+	}
+	if ((*pnt)[0] > _ur[0] || (*pnt)[1] > _ur[1] || (*pnt)[2] > _ur[2]) {
+		ret_pnt = nullptr;
+		return false;
+	}
+
+	// at this place it holds true that the point is within [_ll, _ur]
+	if (!_is_leaf) {
+		for (std::size_t k(0); k < 8; ++k) {
+			if (_childs[k]->addPoint_(pnt, ret_pnt)) {
+				return true;
+			} else {
+				if (ret_pnt != nullptr)
+					return false;
+			}
+		}
+	}
+
+	ret_pnt = pnt;
+	if (_pnts.size() < OctTree<POINT>::_max_items_per_node) {
+		_pnts.push_back(pnt);
+	} else { // i.e. _pnts.size () == OctTree<POINT>::_max_items_per_node
+		splitNode(pnt);
+		_pnts.clear();
+	}
+	return true;
+}
+
+template <typename POINT>
+bool OctTree<POINT>::addPointToChild(POINT * pnt)
+{
+	if ((*pnt)[0] < _ll[0] || (*pnt)[1] < _ll[1] || (*pnt)[2] < _ll[2])
+		return false;
+	if ((*pnt)[0] > _ur[0] || (*pnt)[1] > _ur[1] || (*pnt)[2] > _ur[2])
+		return false;
+
+	if (_pnts.size() < OctTree<POINT>::_max_items_per_node) {
+		_pnts.push_back(pnt);
+	} else { // i.e. _pnts.size () == OctTree<POINT>::_max_items_per_node
+		splitNode(pnt);
```

Comment:
With the current interface this would not compile. 
  1. The OctTree will  only manage pointers to points.
  2. It is only possible to add pointer to points one after another.
  3. Inserting equal points (equal except for eps radius) is not possible, the second occurence will be rejected via a returned false (see addPoint())